Project-Hotpot/Project-Hotpot/Parse-Backend/PFPlaylist.swift

```diff
@@ -33,34 +34,66 @@ class PFPlaylist: PFObject, PFSubclassing {
         })
     }
     
-    class func addSongToPlaylistInBackground(song: PFSong,
-                                             playlist: PFPlaylist,
-                                             completion: @escaping((Bool, Error?) -> Void)) {
-        let relation = playlist.relation(forKey: "songs")
-        relation.add(song)
-        playlist.saveInBackground {success, error in
-            if success == true {
+    
+    class func addSongtoPlaylistInBackground(song: PFSong, playlist: PFPlaylist, completion: @escaping ((Bool, Error?)-> Void)){
+        // create an entry in the Follow table
+        let joinTable = PFObject(className: "SongJoinTable")
+        joinTable.setObject(song, forKey: "Song")
+        joinTable.setObject(playlist, forKey: "Playlist")
+        
+        let currentDate = NSDate()
+        joinTable.setObject(currentDate, forKey: "AddedAt")
+        joinTable.setObject(currentDate, forKey: "LastPlayed")
+        
+        joinTable.saveInBackground { saved, error in
+            if saved {
+                NSLog("song saved successfully")
                 completion(true, nil)
             }
-            else{
+            else {
+                NSLog("failed to save song to playlist")
                 completion(false, error)
             }
         }
     }
     
-    //TODO: change to getting ALL playlists instead of just first.
-    class func getPlaylistInBackground(completion: @escaping(PFPlaylist?, Error?) -> Void) {
+    class func getAllSongsFromPlaylist(playlist: PFPlaylist, completion: @escaping ([PFSong]?, Error?)->Void){
+        let query = PFQuery(className: "SongJoinTable")
+        var songArray: [PFSong]? = []
+        //playlist is a pointer to playlist object
+        query.includeKey("Song")
+        query.whereKey("Playlist", equalTo: playlist)
+        query.order(byAscending: "AddedAt")
+        query.findObjectsInBackground { objects, error in
+            //TODO: is there a more efficient way to do this? similar to O(n) efficiency.
+            if let objects = objects{
+                for o in objects{
+                    if let songObject = o.object(forKey: "Song") as? PFSong{
+                        songArray?.append(songObject)
+                    }
+                }
```

Comment:
You can use `filter`: https://developer.apple.com/documentation/swift/sequence/filter(_:)

Reply:
From my understanding, filter would add the entire query object element to the array, whereas I want specifically the song element stored within the "song" key. I am also attempting to user the "map" function, but am hitting type errors.

Reply:
Ahh you're totally right, my bad! Try `compactMap` in this case: https://developer.apple.com/documentation/swift/sequence/compactmap(_:)

Do you have details on the type errors you're getting?